### PR TITLE
Create new resource type: Microsoft.ResourceGraph/queries

### PR DESCRIPTION
**Pull request by ericc1103**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by adxsdknet** [comment]
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/5593 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/5593/commits/238769ca28e139d41c76d069fefd958694977f71

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]
New swagger should have an entry in https://github.com/Azure/azure-rest-api-specs/blob/master/specification/resourcegraph/resource-manager/readme.md

**Comment by ericc1103** [comment]
@jhendrixMSFT Could you point me the documentation on how to map newly created files to readme?

**Comment by jhendrixMSFT** [comment]
@ericc1103 it might not be documented, I'm following up on that.  Take a look at the other examples in the file, we can iterate on it in this PR.

**Comment by ericc1103** [comment]
@jhendrixMSFT  Updated readme.md and it passes all tests.

**Comment by chiragg4u** [comment]
@ericc1103 please review and get the approval internally in the team before publishing the swagger into the public REPO.

**Comment by ericc1103** [comment]
`please review and get the approval internally in the team before publishing the swagger into the public REPO.`

@chiragg4u What do you mean by that? How do I get an approval internally? From my past experiece, someone in ARM team contacted me to review it. Has the review process changed?

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json
Comment (majastrz) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json]: >201 [](start = 11, length = 3)

As per [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource), this should be 200.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json
Comment (majastrz) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json]: This is redundant since resourcegraph.json already has it.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json
Comment (majastrz) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json]: >Resource [](start = 5, length = 8)

There are common definitions you can reuse via relative paths. Have you tried doing that?

**Review by majastrz — CHANGES_REQUESTED** [review]
Let's wait for the internal review first.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json]: The resorucegraph.json has a lot more properties than it needs so we decided to use a slim version.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json]: This version has less properties than common one.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/savedquery.json
Comment (chiragg4u) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/savedquery.json]: Are we supporting extended to start with?

**Comment by ericc1103** [comment]
@chiragg4u I think I have updated Swagger based on your feedback (version, name, and etc) So would you please take a look at it again and remove `WaitForInternalTeamReview` tag? Please let me know if you want me to address other issues.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/examples/OperationsList.json
Comment (chiragg4u) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/examples/OperationsList.json]: Suggest removing reference to saved word. It's implicit as we are storing as the object. We can add description that it can be used to get the results instead of passing the query at run time.

File: specification/resourcegraph/resource-manager/readme.md
Comment (chiragg4u) [inline specification/resourcegraph/resource-manager/readme.md]: Since we have a separate swagger for the queries; How will it be included in the SDK generations?

**Review by chiragg4u — COMMENTED** [review]
Overall looks good.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json]: I think we have to create new swagger file since this is new resource type.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/examples/OperationsList.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/examples/OperationsList.json]: I think you are looking at the old file. It is already modified and fixed.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/savedquery.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/savedquery.json]: We are not supporting `extended` yet. Will remove it.

File: specification/resourcegraph/resource-manager/readme.md
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/readme.md]: @chiragg4u Do you have any suggestion on this?

**Comment by ericc1103** [comment]
> Let's wait for the internal review first.

@majastrz the internal review has been completed. Could we move forward?

**Comment by majastrz** [comment]
> > Let's wait for the internal review first.
> 
> @majastrz the internal review has been completed. Could we move forward?

The ARM API on call should be able to help with that.

**Comment by ericc1103** [comment]
> > > Let's wait for the internal review first.
> > 
> > 
> > @majastrz the internal review has been completed. Could we move forward?
> 
> The ARM API on call should be able to help with that.

@majastrz could you please at least remove the tag you added? I think I have addressed your comments.

**Comment by ericc1103** [comment]
@KrisBash Could you please take a look at my change and kindly remove the tag if it meets your request?

File: specification/resourcegraph/resource-manager/readme.md
Comment (chiragg4u) [inline specification/resourcegraph/resource-manager/readme.md]: I'm not sure but should we refer to 2019-04-01 for this? Not sure how the SDK's work with different versions.

**Review by chiragg4u — APPROVED** [review]
Approved.

File: specification/resourcegraph/resource-manager/readme.md
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/readme.md]: We have both 2019-04-01 and 2018-09-01 since we still need to support the old version.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (majastrz) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: >null [](start = 26, length = 4)

If you ever want to introduce paging in the future, you would have to introduce a new API version and you'll have constrain the number of results you return on the old API. My suggestion would be to set this to "nextLink" even if you are returning one page currently. That way you you can introduce paging later without breaking clients.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json
Comment (majastrz) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2019-03-01-preview/savedquery.json]: It's not required to have separate swagger per resource type. It mostly lets you rev API versions separately per resource type. We're doing it this way in Blueprint.

---
In reply to: [276460279](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r276460279) [](ancestors = 276460279)

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (majastrz) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: >timeModified [](start = 9, length = 12)

Should this be a string or DateTime in the C# SDK? If it's the latter, you should add "format": "datetime"?

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (majastrz) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: >basic [](start = 13, length = 5)

This enum doesn't have the "extended" value like the one below. Is that intentional?

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (majastrz) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: Btw in general, my suggestion would be to run autorest to generate one of the SDKs (c#, for example) and verify that what it generated seems reasonable.

---
In reply to: [278378779](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r278378779) [](ancestors = 278378779)

**Review by majastrz — CHANGES_REQUESTED** [review]
Mostly looks ok. I added a couple of comments (ignore the old ones just look at the ones from 4/24 or newer).

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (sanjaiganesh) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: Could you confirm whether PUT completes synchronously ? 
I see all operations are marked sync. please review and confirm

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (sanjaiganesh) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: +1

---
In reply to: [278377348](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r278377348) [](ancestors = 278377348)

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (sanjaiganesh) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: >"nextLinkName": null [](start = 10, length = 20)

same feedback here, to specify support for paging

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (sanjaiganesh) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: >Operation [](start = 5, length = 9)

"isDataAction" is missing

IsDataAction:  Required. Indicates whether the operation applies to data-plane. Set the value to "true" for data-plane operations and "false" for ARM/control-plane operations.

**Comment by sanjaiganesh** [comment]
>         "value": [

As a note: all resource providers would need to include the "{Resource Provider Namespace}/register/action" operation in their response. This API is used to register for their service, and should include details about the operation (e.g. a localized name for the resource provider + any special considerations like PII release). Example values can be seen below:
Resource: "Storage Resource Provider"
Operation: "Registers the Storage Resource Provider"
Description: "Registers the subscription for the storage resource provider and enables the creation of storage accounts."

---
Refers to: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/examples/OperationsList.json:8 in 2824f34. [](commit_id = 2824f349226d22bada62c59c9565a98fb242093f, deletion_comment = False)

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: Yes it is

**Comment by ericc1103** [comment]
> > ```
> >     "value": [
> > ```
> 
> As a note: all resource providers would need to include the "{Resource Provider Namespace}/register/action" operation in their response. This API is used to register for their service, and should include details about the operation (e.g. a localized name for the resource provider + any special considerations like PII release). Example values can be seen below:
> Resource: "Storage Resource Provider"
> Operation: "Registers the Storage Resource Provider"
> Description: "Registers the subscription for the storage resource provider and enables the creation of storage accounts."
> 
> Refers to: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/examples/OperationsList.json:8 in [2824f34](https://github.com/Azure/azure-rest-api-specs/commit/2824f349226d22bada62c59c9565a98fb242093f). [](commit_id = [2824f34](https://github.com/Azure/azure-rest-api-specs/commit/2824f349226d22bada62c59c9565a98fb242093f), deletion_comment = False)

This resource doesn't require to be registered. We have marked it as `RegistrationFree`

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: We are not going to support Extended yet. It will come later when we enable that feature.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: Not sure why we need this field. Could you elaborate more?

**Comment by ericc1103** [comment]
@sanjaiganesh & @majastrz, I have addressed the most of comments, please validate those and let me know.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (majastrz) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: NextLink is still null here.

---
In reply to: [278412814](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r278412814) [](ancestors = 278412814,278377348)

Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: Just updated as well.

Comment (majastrz) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: See it now. Thanks!

---
In reply to: [279135228](https://github.com/Azure/azure-rest-api-specs/pull/5478#discussion_r279135228) [](ancestors = 279135228,278412814,278377348)

**Review by majastrz — APPROVED** [review]
Looks good.

**Comment by ericc1103** [comment]
@jhendrixMSFT, would you please remove the `requested changes` flag if everything looks OK?

**Review by gauravkapila — APPROVED** [review]

**Comment by ericc1103** [comment]
@gauravkapila Thanks for the approval. But the `requested changes` should be removed that @jhendrixMSFT created. I have addressed his request so please remove it since you are DRI.

**Review by jhendrixMSFT — APPROVED** [review]
Waiting for CI to finish, will merge assuming everything is good.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (jhendrixMSFT) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: This is missing the `nextLink` field as defined in the `x-ms-pageable` extension for corresponding operation.

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: Thanks for catching this! Just updated.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (jhendrixMSFT) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: This same path is defined in resourcegraph.json which is included in this same package, and since the return types are slightly different (nextLinkName is null in the other copy) it causes problems.  Can you please remove one of this?

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json
Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: @chiragg4u Do you want me to go ahead and fix this or want to handle it separately since it is not really a scope of my PR?

Comment (jhendrixMSFT) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: It needs to be fixed, without it we cannot generate SDKs.

Comment (yangmsft) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: @jhendrixMSFT Thanks for the heads-up. We will fix resourcegraph.json in a separate PR.

Comment (jhendrixMSFT) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: Are you going to remove the duplicate entry from graphquery.json then?

Comment (jhendrixMSFT) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: Sorry if I wasn't clear, the build for this new tag `package-2018-09-preview` is broken and must be fixed before we can merge this (see red CI).  You can either remove the duplicate entry added in graphquery.json or fix resourcegraph.json so the definitions are identical (I don't know which fix is correct so you will have to decide).

Comment (ericc1103) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/preview/2018-09-01-preview/graphquery.json]: Just removed but seeing other issue and looking into it.

**Review by jhendrixMSFT — APPROVED** [review]

**Pull request merged**
